hack/actions/check-changefile-exists.go

```diff
@@ -0,0 +1,136 @@
+	// and allows multiline strings to work properly.
+	log.SetFormatter(&logrus.TextFormatter{
+		ForceColors: true,
+	})
+	// We need a GITHUB_TOKEN and PR_NUMBER in the environment.
+	// These are set by the Action config file
+	// in .github/workflows/prbuild.yaml,
+	// under the check-changelog step.
+	token, ok := os.LookupEnv("GITHUB_TOKEN")
+	if !ok {
+		log.Fatal("No GITHUB_TOKEN set, check the Action config.")
+
+	}
+	prEnv, ok := os.LookupEnv("PR_NUMBER")
+	if !ok {
+		log.Fatal("No PR_NUMBER set, check the Action config.")
+	}
+	pr, err := strconv.Atoi(prEnv)
+	if err != nil {
+		log.Fatalf("Couldn't convert PR number, %s", err)
+	}
+
+	// We've got what we need, set up the Github client to get the
+	// labels associated with the PR.
+	ctx := context.Background()
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+
+	client := github.NewClient(tc)
+
+	prDetails, _, err := client.PullRequests.Get(ctx, "projectcontour", "contour", pr)
+	if err != nil {
+		log.Fatalf("Couldn't get PR details: %s", err)
+	}
+
+	// No labels. This is most likely what people will see at first, so this message
+	// is as friendly as I could make it.
+	if len(prDetails.Labels) == 0 {
+		log.Fatal(`
+Thanks for your PR.
+For a PR to be accepted to Contour, it must have:
+- at least one release-note label set
+- a file named changelogs/unreleased/PR#-author-category,
+	where category matches the relase-note/category label you apply`)
+
+	}
+
+	// Try to determine the category of the PR.
+	var category string
+	for _, label := range prDetails.Labels {
+		name := *label.Name
+		if strings.HasPrefix(name, "release-note") {
+			// In case the old release-note labels stick around, mark them
+			// as "major" category.
+			if name == "release-note" || name == "release-note-action-required" {
+				category = "major"
+				break
+			}
+
+			// Otherwise, extract the category.
+			labelSplit := strings.Split(name, "/")
+			if len(labelSplit) > 1 {
+				category = labelSplit[1]
+			}
+		}
+	}
+
+	if category == "" {
+		log.Fatal(`
+Thanks for your PR.
+For a PR to be accepted to Contour, it must have:
+- at least one release-note label set
+- a file named changelogs/unreleased/PR#-author-category,
+  where category matches the relase-note/category label you apply.
+
+There are some labels set, but there must be at least one release-note label.`)
+	}
+
+	// None required is the escape hatch for small changes.
+	if category == "none-required" {
+		log.Println("No changelog required.")
+		os.Exit(0)
+	}
+
+	changelogFile, err := os.Stat(fmt.Sprintf("./changelogs/unreleased/%d-%s-%s.md",
+		pr, *prDetails.User.Login, category))
+
+	if os.IsNotExist(err) {
+		log.Fatalf(`
+Thanks for your PR, and thanks for labelling it with a release-note.
+For a PR to be accepted to Contour, it must have a file named
+changelogs/unreleased/%d-%s-%s.md with a description of the change.`,
```

Comment:
Just some general questions and thoughts about the changelog files:

I'm guessing that category is part of the changelog filename to make it easier to collect the entries under certain heading in the release notes? And the reason for user name in filename is for giving shoutout for the contributor in the release notes?

In theory this info could be also collected at the time of release from the github PRs. Category can possibly change during development of the PR, and I assume user may change the label even after the check was already run. It might also happen more often than not, that the used template file was wrong, causing e.g. formatting problems with wrong heading levels.

There probably is many projects using automatic changelogs, but I've seen similar process in Vault where the changelog entries used format like this https://github.com/hashicorp/vault/ including some metadata in the document itself, collected into sections like this https://github.com/hashicorp/vault/blob/main/CHANGELOG.md. Though that seems bit more limited, allowing only one-liners and not longer descriptions / formatting.

Reply:
Yes, right now we have very detailed changelogs, and that's something I definitely wanted to keep, so I wanted to have a mechanism to be able to have a reasonably large amount of explanation for any particular change.

I lifted a lot of this design from Velero and Envoy, as it seems to be working reasonably well for them.

I agree that the downside is that, best case, you have to do a second ccommit with the changelog for your PR once you know the number, but I think the benefits we will get will be worth it. In particular, I would like for it not to take a full day of writing release notes before we can release, while hopefully keeping the changelogs to the standard we've been delivering.

Reply:
Yeah, having to do the second commit is a minor nuisance for the benefit of not requiring a full work-day for release notes!

As an alternative to putting "metadata" like `category` and `username` in filename, there could be release tooling that fetches that info on-demand when writing release notes.  Username of course does not change during the life-cycle of the PR, but category might change. 

But same as in my other comment: this is just a minor thing that came to my mind, thinking what could go wrong in practice :)